File: testdata/py_fixture/store.py
# Realistic Python fixture for resolver measurement.
# Exercises: annotated class fields, __init__ self-assignment, typed params,
# self.field.method(), self.method(), constructor-typed locals.


class Conn:
    def query(self, sql):
        return []

    def close(self):
        pass


class Row:
    def value(self):
        return ""


class Repository:
    conn: Conn

    def __init__(self):
        self.cache = Conn()

    def load(self, id):
        self.conn.query("select")   # annotated field -> Conn.query
        self.cache.query("select")  # __init__ assignment -> Conn.query
        self.cleanup()              # self.method -> Repository.cleanup

    def cleanup(self):
        self.conn.close()           # annotated field -> Conn.close


class Service:
    def run(self, c: Conn):
        c.query("x")     # typed param -> Conn.query
        c.close()        # typed param -> Conn.close
        local = Conn()
        local.query("y") # constructor-typed local -> Conn.query

File: testdata/ts_fixture/store.ts
// Realistic TypeScript fixture for resolver measurement.
// Exercises: field types, typed params, return types, this.field.m(),
// this.m(), new X() locals, return-typed locals.

export class Conn {
  query(sql: string): Row[] {
    return [];
  }
  close(): void {}
}

export class Row {
  value(): string {
    return "";
  }
}

export class Repository {
  private conn: Conn;

  constructor(conn: Conn) {
    this.conn = conn;
  }

  load(id: string): Row[] {
    const rows = this.conn.query("select"); // field receiver -> Conn.query
    this.cleanup();                          // this.method -> Repository.cleanup
    return rows;
  }

  cleanup(): void {
    this.conn.close(); // field receiver -> Conn.close
  }
}

function newConn(): Conn {
  return new Conn();
}

export class Service {
  private repo: Repository;

  run(id: string): void {
    const c = newConn(); // return-typed local -> Conn
    c.query("x");        // -> Conn.query
    const fresh = new Conn(); // new local -> Conn
    fresh.close();            // -> Conn.close
    this.repo.load(id);       // field receiver -> Repository.load
  }
}
